lib/filewriter.js

```diff
@@ -0,0 +1,182 @@
+    this.write = function(str, cb) {
+        var self = this;
+        this._getLockedFd(function(err, fd) {
+            if (err) return cb(err);
+            // not much benefit to reusing a buffer for the writes
+            var buf = new Buffer(str);
+            fs.write(fd, buf, 0, buf.length, null, function(err, nb) {
+                flockSync(fd, "un");
+                cb(err, nb);
+            });
+        });
+    };
+
+    // close the file
+    this.close = function( ) {
+        try { if (this.fd) fs.closeSync(this.fd); }
+        catch (err) { }
+        this.fd = undefined;
+    };
+
+    // expose class method on instance
+    this.mutexTimeout = FileWriter.mutexTimeout;
+    this.renameFile = FileWriter.renameFile;
+}
+
+
+// rename the file and wait for write activity to cease
+FileWriter.mutexTimeout = 5000;
+FileWriter.renameFile = function renameFile( oldName, newName, waitMs, cb ) {
+    if (!cb && typeof waitMs === 'function') {
+        cb = waitMs
+        waitMs = 50
+    }
+    // Wait for all pending writes to finish.  After waitMs no writer will reuse the old fd,
+    // and once we`ve paused and have the mutex we know that no more writes will occur.
+    var fd = null, mutexTimeout = (this && this.mutexTimeout) ? this.mutexTimeout : FileWriter.mutexTimeout;
+    function waitForWritesToFinish( ) {
+        var guard, error = null
+        function finish(err) {
+            if (fd !== null) fs.close(fd, function(){})
+            fd = null
+            if (guard && global.clearTimeout) clearTimeout(guard)
+            guard = null
+            return error ? cb(error) : cb(err)
+        }
+        try {
+            guard = setTimeout(function() {
+                if (fd !== null) fs.close(fd, function(){})
+                error = new Error("timed out waiting for last write")
+                return finish(error)
+            }, mutexTimeout)
+            if (guard.unref) guard.unref()
+            fd = fs.openSync(newName, 'r')
+            fse.flock(fd, 'ex', function() {
+                try {
+                    fse.flockSync(fd, 'un')
+                    fs.closeSync(fd)
+                    if (!error) return finish()
+                }
+                catch (err) {
+                    if (!error) return finish(err)
+                }
+            })
+        }
+        catch (err) {
+            return finish(err)
+        }
+    }
+    try {
+        // only rename if the source exists, else an empty target is created "wx"
+        var rfd = fs.openSync(oldName, "r")
+        fs.closeSync(rfd)
+        // renameSync overwrites an existing newName, does not throw an error
+        // prevent a rename-rename race condition by first getting exclusive rights to newName
+        // note: race condition: if the source disappears, an empty newName can be created
+        var fd = fs.openSync(newName, 'wx')
+        fs.closeSync(fd)
+        fs.renameSync(oldName, newName)
```

Comment:
Do we need to unset the ex lock after the rename, or is that done automatically?